tests/test_resources/test_modules/test_functions/test_function.py

```diff
@@ -173,301 +75,469 @@ async def async_slow_generator(size):
+        re_fn = rh.function(nb_sum).to(cluster, env=["torch==1.12.1"])
+
+        re_fn.notebook()
+        summands = list(zip(range(5), range(4, 9)))
+        res = re_fn(summands)
+
+        assert res == [4, 6, 8, 10, 12]
+        re_fn.delete_configs()
+
+    @unittest.skip("Runs indefinitely.")
+    def test_ssh(self):
+        # TODO do this properly
+        my_function = rh.function(name="local_function")
+        my_function.ssh()
+        assert True
+
+    @pytest.mark.rnstest
+    @pytest.mark.level("local")
+    def test_share_function(self, cluster):
+        my_function = rh.function(fn=summer).to(cluster).save(REMOTE_FUNC_NAME)
+
+        my_function.share(
+            users=["donny@run.house", "josh@run.house"],
+            access_type="read",
+            notify_users=False,
+        )
+        assert True
+
+    @pytest.mark.rnstest
+    @pytest.mark.level("local")
+    def test_load_shared_function(self):
+        my_function = rh.function(name=REMOTE_FUNC_NAME)
+        res = my_function(1, 2)
+        assert res == 3
+
+    @pytest.mark.rnstest
+    @pytest.mark.level("local")
+    def delete_function_from_rns(self, s):
+        server_url = s.rns_client.api_server_url
+        resource_request_uri = s.rns_client.resource_uri(s.name)
+        resp = requests.delete(
+            f"{server_url}/resource/{resource_request_uri}",
+            headers=s.rns_client.request_headers,
+        )
+        if resp.status_code != 200:
+            raise Exception(
+                f"Failed to delete_configs function data from path: {load_resp_content(resp)}"
+            )
+
+        try:
+            # Terminate the cluster
+            s.cluster.teardown_and_delete()
+        except Exception as e:
+            raise Exception(f"Failed to teardown the cluster: {e}")
+
+    @pytest.mark.clustertest
+    @pytest.mark.rnstest
+    @pytest.mark.level("thorough")
+    def test_load_function_in_new_env(self, ondemand_cpu_cluster, byo_cpu):
+        ondemand_cpu_cluster.save(
+            f"@/{ondemand_cpu_cluster.name}"
+        )  # Needs to be saved to rns, right now has a local name by default
+        remote_sum = rh.function(summer).to(ondemand_cpu_cluster).save(REMOTE_FUNC_NAME)
+
+        remote_python = (
+            "import runhouse as rh; "
+            f"remote_sum = rh.function(name='{REMOTE_FUNC_NAME}'); "
+            "res = remote_sum(1, 5); "
+            "assert res == 6"
+        )
+        res = byo_cpu.run_python([remote_python], stream_logs=True)
+        assert res[0][0] == 0
+
+        remote_sum.delete_configs()
+
+    @pytest.mark.clustertest
+    @pytest.mark.level("thorough")
+    def test_nested_diff_clusters(self, ondemand_cpu_cluster, byo_cpu):
+        summer_cpu = rh.function(summer).to(ondemand_cpu_cluster)
+        call_function_diff_cpu = rh.function(call_function).to(byo_cpu)
+
+        kwargs = {"a": 1, "b": 5}
+        res = call_function_diff_cpu(summer_cpu, **kwargs)
+        assert res == 6
```

Comment:
These two also need a significant update. Same with the http_url ones below.

Reply:
marked to refactor those tests. (TODO)